File: LeetCode/Patterns/Trees/98. Validate Binary Search Tree/Solution.java
class Solution {
    public boolean isValidBST(TreeNode root) {
        return validate(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean validate(TreeNode node, long min, long max) {
        if (node == null) return true;

        // Current node must lie strictly between min and max
        if (node.val <= min || node.val >= max) {
            return false;
        }

        // Check left and right subtrees with updated bounds
        return validate(node.left, min, node.val) &&
               validate(node.right, node.val, max);
    }
}

File: LeetCode/Patterns/Trees/98. Validate Binary Search Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* prev = nullptr;
    bool ans = true;

    void inorder(TreeNode* root){
        if(root == nullptr)
          return;
        if(prev != nullptr && prev->val >= root->val){
            ans = false;
            return;
        }
        prev = root;
        inorder(root->right);
    }
    bool isValidBST(TreeNode* root) {
        inorder(root);
        return ans;
        
    }
};